### PR TITLE
Fixed #653: Handle Multiple emergency contacts

### LINKED ISSUE #653: Handle Multiple emergency contacts
### Expected Behaviour

A client should be able to have multiple emergency contacts attached to him. The creation/edition form must allow multiple selection, and the client's view should show all related contacts.

### Actual Behaviour

Only one contact is supported per client.

### Steps to reproduce

* Find a client with multiple emergency contacts 
* Edit client 
* Click emergency tabs 
* ...

see screenshot 
multipleemergencycontact

### PR DESCRIPTION
**Pull request by aaboffill**
### Changes proposed in this pull request:

* Changed the `emergency contact` relation to `ManyToManyField`. 
* Changed the corresponding view to use `django formset`
* Fixed all related tests.

### Status

- [X] READY
- [ ] HOLD
- [ ] WIP (Work-In-Progress)

### How to verify this change

* Create or modify a `Client`.
* Go to the `Emergency` tab.
* Add several `Emergency contacts` 

### Deployment notes and migration

- [X] Migration is needed for this change

### New translatable strings

- [X] If applicable, I have included updated .po files with new strings (see http://goo.gl/kfBO7N)

### CODE REVIEW

File: src/member/migrations/0026_change_to_emergency_contacts.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='member.Member',
+                                    verbose_name='member'),
+        ),
+        migrations.AddField(
+            model_name='client',
+            name='emergency_contacts',
+            field=models.ManyToManyField(related_name='emergency_contacts_of', through='member.EmergencyContact',
+                                         to='member.Member', verbose_name='emergency contacts'),
+        ),
+        migrations.RunPython(migrate_emergency_contacts),
```
Comment (lingxiaoyang) [inline src/member/migrations/0026_change_to_emergency_contacts.py]: For `RunPython`, we may need to provide `reverse_code` to unapply the changes for testing purposes, and to safely redo this migration:
https://docs.djangoproject.com/en/1.10/ref/migration-operations/#runpython

File: src/member/templates/client/view/information.html [hunk 2]
```diff
@@ -52,32 +52,35 @@ <h2 class="ui header">
 </div>
 
 
-{% if client.emergency_contact %}
+{% if client.emergencycontact_set.all|length %}
```
Comment (lingxiaoyang) [inline src/member/templates/client/view/information.html]: I think we can use `emergencycontact_set.exists`, if I'm correct.

File: src/member/templates/client/view/information.html [hunk 3]
```diff
@@ -52,32 +52,35 @@ <h2 class="ui header">
...
 <div class="ui segment">
     <h2 class="ui header">
         <div class="content">
-            {% trans 'Emergency Contact' %}
-            <div class="sub header">{% trans "Information of the person to contact in case of an emergency" %}.</div>
+            {% trans 'Emergency Contacts' %}
+            <div class="sub header">{% trans "Information of the persons to contact in case of an emergency" %}.</div>
```
Comment (lingxiaoyang) [inline src/member/templates/client/view/information.html]: persons -> people

File: src/member/migrations/0026_change_to_emergency_contacts.py (hunk 1 above)
Comment (aaboffill) [inline src/member/migrations/0026_change_to_emergency_contacts.py]: Nice! @lingxiaoyang ... no problem, that's is important, i'll add the reverse function.

File: src/member/templates/client/view/information.html (hunk 2 above)
Comment (aaboffill) [inline src/member/templates/client/view/information.html]: I'll test that, in theory should work fine your proposal.

File: src/member/models.py [hunk 4]
```diff
@@ -1034,6 +1027,25 @@ def __str__(self):
...
+    relationship = models.CharField(
+        max_length=100,
+        verbose_name=_('relationship'),
+        blank=True,
+        null=True,
+    )
+
+    class Meta:
+        verbose_name_plural = _('emergency contacts')
+
```
Comment (lingxiaoyang) [inline src/member/models.py]: There could be a `unique_together = ('client', 'member')`.

**Comment by aaboffill** [comment]
I think that with 'ui raised segment' the UI has improved a lot. About the buttons (remove/add), by the moment I think that simple links is better to keep the page clear.

**Comment by lingxiaoyang** [comment]
@aaboffill  Yes I agree. If we use buttons, there will be too many colors on the same page.

File: src/member/migrations/0026_change_to_emergency_contacts.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    Client = apps.get_model('member', 'Client')
+    for client in Client.objects.all():
+        # Use as client emergency contact the first item of
+        # relationship
+        emergency_contact = client.emergencycontact_set.first()
+        # reversing info
+        client.emergency_contact = emergency_contact.member
+        client.emergency_contact_relationship = emergency_contact.relationship
+        # removing info
+        client.emergencycontact_set.delete()
```
Comment (lingxiaoyang) [inline src/member/migrations/0026_change_to_emergency_contacts.py]: @aaboffill  An error here:
```python
  File "/code/src/member/migrations/0026_change_to_emergency_contacts.py", line 33, in reverse_emergency_contacts
    client.emergencycontact_set.delete()
AttributeError: 'RelatedManager' object has no attribute 'delete'
```

Use `client.emergencycontact_set.clear()`.

File: src/member/migrations/0027_auto_20170221_1438.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    dependencies = [
+        ('member', '0026_change_to_emergency_contacts'),
+    ]
+
+    operations = [
+        migrations.AlterUniqueTogether(
+            name='emergencycontact',
+            unique_together=set([('client', 'member')]),
+        ),
+    ]
```
Comment (lingxiaoyang) [inline src/member/migrations/0027_auto_20170221_1438.py]: This file can be merged into 0026.

**Comment by lingxiaoyang** [comment]
@aaboffill  Thank you! Two more comments on fef869b:

1. While creating a new client, if I leave the emergency contact form as blank and click "save", it returns a 500 page. This could be more user-friendly.
image
image

2. It could be the case that a client has no emergency contact in the DB. When editing these clients, the form is not displaying:
image

**Comment by lingxiaoyang** [comment]
The 'ui raised segment' is really nice!

File: src/member/migrations/0026_change_to_emergency_contacts.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        ),
+        migrations.RemoveField(
+            model_name='client',
+            name='emergency_contact_relationship',
+        ),
+        migrations.AlterField(
+            model_name='route',
+            name='vehicle',
+            field=models.CharField(choices=[('cycling', 'Cycling'), ('walking', 'Walking'), ('driving', 'Driving')], default='cycling', max_length=20, verbose_name='vehicle'),
+        ),
```
Comment (lingxiaoyang) [inline src/member/migrations/0026_change_to_emergency_contacts.py]: @aaboffill  This is addressed in #670.

File: src/member/migrations/0026_change_to_emergency_contacts.py (hunk 5 above)
Comment (aaboffill) [inline src/member/migrations/0026_change_to_emergency_contacts.py]: I've the same problem with `clear` method, finnaly I changed `client.emergencycontact_set.delete()` by `EmergencyContact.objects.filter(client=client).delete()`

File: src/member/migrations/0026_change_to_emergency_contacts.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            )
+
+
+def reverse_emergency_contacts(apps, schema_editor):
+    # set NotePriority initial
+    Client = apps.get_model('member', 'Client')
+    for client in Client.objects.all():
+        # Use as client emergency contact the first item of
+        # relationship
+        emergency_contact = client.emergencycontact_set.first()
```
Comment (aaboffill) [inline src/member/migrations/0026_change_to_emergency_contacts.py]: I agree...

File: src/member/migrations/0027_auto_20170221_1438.py (hunk 6 above)
Comment (aaboffill) [inline src/member/migrations/0027_auto_20170221_1438.py]: I agree, in production the migration 0026 is not applied, is better to merge both migrations.

**Comment by aaboffill** [comment]
@lingxiaoyang I answered the last comments...

File: src/member/migrations/0027_auto_20170221_1438.py (hunk 6 above)
Comment (lingxiaoyang) [inline src/member/migrations/0027_auto_20170221_1438.py]: We can remove this file then. 0026 now contains the same operation.

**Comment by aaboffill** [comment]
ohhh!! I forgot to remove the file '0027_auto_20170221_1438.py' :)

File: src/member/migrations/0026_change_to_emergency_contacts.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    EmergencyContact = apps.get_model('member', 'EmergencyContact')
+    for client in Client.objects.all():
+        if client.emergencycontact_set.exists():
+            # Use as client emergency contact the first item of
+            # relationship
+            emergency_contact = client.emergencycontact_set.first()
+            # reversing info
+            client.emergency_contact = emergency_contact.member
+            client.emergency_contact_relationship = emergency_contact.relationship
+            # removing info
```
Comment (lingxiaoyang) [inline src/member/migrations/0026_change_to_emergency_contacts.py]: You forgot `client.save()`!

Sorry I didn't find it earlier...

Comment (aaboffill) [inline src/member/migrations/0026_change_to_emergency_contacts.py]: uff done!!!

**Comment by aaboffill** [comment]
@lingxiaoyang the confilct has been resolved...

**Comment by lingxiaoyang** [comment]
@erozqba  If you have time tomorrow, could you also review/test this PR, to make sure it doesn't bring any problems to the production environment?

**Comment by erozqba** [comment]
It seems that the add and remove link are still broken. 

image

**Comment by lingxiaoyang** [comment]
@erozqba  Have you run "gulp" after merging the PR? There are some new JS codes.

**Comment by erozqba** [comment]
@lingxiaoyang yes I had run gulp, but I will do it again only to be sure!

**Comment by lingxiaoyang** [comment]
@erozqba  OK!

**Comment by erozqba** [comment]
:+1: @lingxiaoyang Its works as expected! 
But I don't have any production data so It will be @manumilou that can test this with production data and tell us if there is any problem.

**Comment by lingxiaoyang** [comment]
@erozqba Cool! Thank you!

**Comment by erozqba** [comment]
@lingxiaoyang I think is a great idea! Could you do it? Thanks in advance :+1:

**Comment by lingxiaoyang** [comment]
@erozqba  Sure. I'll let you know!

**Comment by manumilou** [comment]
@lingxiaoyang @erozqba Still waiting for me to merge ?

**Comment by lingxiaoyang** [comment]
@manumilou  Yep. You can check out #683 instead of this PR.

**Comment by lingxiaoyang** [comment]
Closing after merging the replacement PR #683

**Pull request closed without merging**